File: LeetCode/binaryTree/297.serialize-and-deserialize-binary-tree.py
#
# @lc app=leetcode id=297 lang=python3
#
# [297] Serialize and Deserialize Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    sequence = ''
    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if(root is None): 
            self.sequence+='#,'       
            return self.sequence
        self.sequence+=str(root.val) +','

        self.serialize(root.left)
        self.serialize(root.right)
        return self.sequence
        

    def deserializeRecur(self, seqList):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """    
        if(not seqList): return None
        rootVal = seqList.pop(0)
        if rootVal == '#': return None 
        # this is the key: return None two time we get a leaf, return None one times we get a single leaf tree
        root = TreeNode(val=rootVal)
        root.left = self.deserializeRecur(seqList)
        root.right = self.deserializeRecur(seqList)

        return root

    def deserialize(self, data):
        print(data)
        seqList = [d for d in data.split(',') if d]
        return self.deserializeRecur(seqList)

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))
# @lc code=end

File: LeetCode_Go/BinaryTree/297.serialize-and-deserialize-binary-tree.go
/*
 * @lc app=leetcode id=297 lang=golang
 *
 * [297] Serialize and Deserialize Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// Encodes a tree to a single string. Implement serialize and deserialize algorithms.

// Preorder Serialization (with null), DFS, append node values or null_string to string builder in preorder place. Deserizalization: The first token in string is the root. DFS and process root node token in preorder place. Then removing the first token before recursive call.
// Serialization: O(n)
// Deserialization: O(n)
import (
	"strconv"
	"strings"
	"log"
)

var SEP byte = ','
var NULL byte = '#'

type Codec struct {
}

func Constructor() Codec {
	return Codec{}
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
	var builder strings.Builder
	this.serializeHelper(root, &builder)
	originalStr := builder.String()
	// fmt.Println(originalStr[:len(originalStr)-1])
	// fmt.Println("serialzation END <<<<<<")
	return originalStr[:len(originalStr)-1] // remove the last comma
}

// preorder DFS and build the string
func (this *Codec) serializeHelper(root *TreeNode, builder *strings.Builder) {

	if root == nil {
		builder.WriteByte(NULL)
		builder.WriteByte(SEP)
		return
	}

	// append current node to builder
	builder.WriteString(strconv.Itoa(root.Val))
	builder.WriteByte(SEP)

	this.serializeHelper(root.Left, builder)
	this.serializeHelper(root.Right, builder)
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
	// turn a string separated by "," to a slice of characters (excluding the ",")
	dataSlice := strings.Split(data, string(SEP))
	// fmt.Println(dataSlice)
	return deserializeHelper(&dataSlice)
}

func deserializeHelper(dataSlice *[]string) *TreeNode {
	if len(*dataSlice) == 0 {
		return nil
	}
	// fmt.Println("current node: ", (*dataSlice)[0])
	if (*dataSlice)[0] == string(NULL) {
		*dataSlice = (*dataSlice)[1:] // remove first element after finishing a node
		// fmt.Println(*dataSlice)
		// fmt.Println("node process END <<<<")
		return nil
	}
	val, err := strconv.Atoi((*dataSlice)[0])
	if err != nil {
		log.Fatal("Error: %v", err)
	}
	root := &TreeNode{Val: val}
	*dataSlice = (*dataSlice)[1:] // remove first element after finishing a node
	// fmt.Println(*dataSlice)
	// fmt.Println("node process END <<<<")

	root.Left = deserializeHelper(dataSlice)
	root.Right = deserializeHelper(dataSlice)

	return root
}

/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor();
 * deser := Constructor();
 * data := ser.serialize(root);
 * ans := deser.deserialize(data);
 */
// @lc code=end

